### ISSUE TITLE
System.load() at build time should persist in runtime or give an error

### ISSUE THREAD

**Issue by CGDogan** [oracle/graal#7015]
System.load() at build time succeeds but the required runtime effect isn't achieved. This is bad in the common case of:

```java
class Image {

static boolean isLoaded = false;

void Image() {
 if (!isLoaded)
  System.load(file);
  isLoaded = true;
 }
}

class Man {
static image = new Image();
}
```

and in runtime it isn't loaded but isLoaded is true

**Comment by kkriske** [comment]
My understanding of how exactly library loading works is incomplete, so I might be off on some things, but:
`System#load` and `System#loadLibrary` load dynamic libraries into memory. This creates runtime specific memory allocations that are impossible to store at build-time. Storing, and using, these memory locations will result in segfaults as you're accessing memory that does not belong to your current execution.

in short, `System#load` and `System#loadLibrary` calls should happen at run-time, not at build-time.

**Comment by CGDogan** [comment]
Yes, hence I'd expect an error at build time

**Comment by kkriske** [comment]
It is still supported to load libraries at build-time as well, as the image builder cannot run without for example the `jvm` and `java` shared libraries.
How such cases are generally caught are with the `rerun` initialization strategy, which is not exposed in the same way as `--initialize-at-build-time=` and `--initialize-at-run-time=` options as it's an expert option that once again when used incorrectly can have major unexpected side effects.
What `rerun` does, is initialize the class at build-time, allowing it to be used by the image builder, but include a non-initialized version of that class in the native-image so the class initializer runs at runtime causing initialization of dynamic libaries to correctly run at runtime as well.

**Comment by CGDogan** [comment]
Oof I guess I'll use rerun then. Thanks for pointing these out, these are much less simpler than I thought.

**Comment by kkriske** [comment]
You shouldn't use `rerun`, it's internal for a good reason.
In your example, you should let the `Image` class initialize at runtime.
